src/main/java/com/hivemq/mqtt/message/pool/FreeIdRanges.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright 2019-present HiveMQ GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hivemq.mqtt.message.pool;
+
+import com.hivemq.extension.sdk.api.annotations.NotNull;
+import com.hivemq.extension.sdk.api.annotations.Nullable;
+import com.hivemq.mqtt.message.pool.exception.NoMessageIdAvailableException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The purpose of this class is to reduce packet IDs allocation time for each independent client
+ * and to reduce the memory footprint of keeping track of allocated IDs.
+ * <p>
+ * This is achieved by keeping a list of {@link Range} objects each of which represents a contiguous interval of
+ * integer ids that are NOT currently assigned to any object using this instance of {@link FreeIdRanges}.
+ * The lower end of the {@link Range} instance is included whereas the upper end is excluded from the interval.
+ * That is, the {@link Range} instance with start 42 and end 42 contains only one ID - 42.
+ * The {@link Range} instance with start 10 and end 12 contains only two IDs: 10 and 11.
+ * <p>
+ * Initially, there is only one contiguous {@link Range} of IDs. The IDs are assigned starting from the lowest one in
+ * that interval.
+ * Upon assignment, the interval's lower end is incremented (interval size reduces from below).
+ * When the ID is returned, it either joins one of the existing {@link Range} intervals in the list (if it is adjacent)
+ * or it forms a new {@link Range} that is added to the list.
+ * <p>
+ * This class is NOT thread-safe.
+ * <a
+ * href="https://github.com/hivemq/hivemq-mqtt-client/blob/master/src/main/java/com/hivemq/client/internal/util/Ranges.java">The
+ * original implementation in the HiveMQ Java Client.</a>
+ */
+public class FreeIdRanges {
+
+    private static final @NotNull Logger log = LoggerFactory.getLogger(FreeIdRanges.class);
+
+    final int minAllowedId;
+    final int maxAllowedId;
```

Comment:
What about hardcoding them? The spec is fixed on 1..65535 for packet IDs anyway.

Reply:
Yes, can be done, the original class was quite generic; making it a bit more specialized (like the original pool) should be fine since it is unlikely that there are other places in the code that profit from it